File: app/src/main/java/android/mahendra/attendancemanager/dialogs/PeriodDialogFragment.java
package android.mahendra.attendancemanager.dialogs;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.mahendra.attendancemanager.R;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.Spinner;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;

import java.util.ArrayList;

public class PeriodDialogFragment extends DialogFragment {
    private static final String TAG = "PeriodDialogFragment";
    private static final String ARG_PERIOD_TITLE = "period title";
    private static final String KEY_TITLE_OPTIONS = "title options";
    private static final String KEY_PERIOD_TITLE = "period_title";

    private String mSubjectTitle;
    private ArrayList<String> mSubjectTitles = new ArrayList<>();
    private Callbacks mCallback;

    public interface Callbacks {
        ArrayList<String> getSubjectTitles();
        void onPeriodSelected(String title);
        void onDeletePeriod(String title);
    }

    public static PeriodDialogFragment newInstance(String periodTitle) {
        Bundle args = new Bundle();
        args.putString(ARG_PERIOD_TITLE, periodTitle);
        PeriodDialogFragment fragment = new PeriodDialogFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString(KEY_PERIOD_TITLE, mSubjectTitle);
        outState.putStringArrayList(KEY_TITLE_OPTIONS, mSubjectTitles);
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        try {
            mCallback = (Callbacks) context;
        }
        catch (ClassCastException ex) {
            Log.e(TAG, "onAttach: " + ex.toString());
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mCallback = null;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (savedInstanceState != null) {
            mSubjectTitle = savedInstanceState.getString(KEY_PERIOD_TITLE, null);
            mSubjectTitles = savedInstanceState.getStringArrayList(KEY_TITLE_OPTIONS);
        }
        else {
            mSubjectTitles = mCallback.getSubjectTitles();
            mSubjectTitle = getArguments().getString(ARG_PERIOD_TITLE, null);
        }
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        Log.i(TAG, "onCreateDialog: created dialog");
        int dialogTitle;
        int positiveButtonText;

        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        LayoutInflater inflater = requireActivity().getLayoutInflater();
        View v = inflater.inflate(R.layout.dialog_choose_period, null);

        Button clearPeriod = v.findViewById(R.id.clear_period);
        clearPeriod.setOnClickListener(v1 -> {
            mCallback.onDeletePeriod(mSubjectTitle);
            Toast.makeText(getActivity(), "Removed " + mSubjectTitle, Toast.LENGTH_SHORT).show();
            dismiss();
        });

        if (mSubjectTitle != null) {
            clearPeriod.setVisibility(View.VISIBLE);
            dialogTitle = R.string.update_period;
            positiveButtonText = R.string.update;
        }
        else {
            clearPeriod.setVisibility(View.GONE);
            dialogTitle = R.string.add_period;
            positiveButtonText = R.string.add;
        }

        Spinner spinner = v.findViewById(R.id.spinner_select_period);
        ArrayAdapter<String> adapter = new ArrayAdapter(getActivity(), android.R.layout.simple_spinner_item);
        adapter.addAll(mSubjectTitles);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinner.setAdapter(adapter);

        builder.setTitle(dialogTitle);
        builder.setView(v);

        if (mSubjectTitle != null) {
            int position = adapter.getPosition(mSubjectTitle);
            spinner.setSelection(position);
        }

        builder.setPositiveButton(positiveButtonText, (dialog, which) -> {
            mCallback.onPeriodSelected(spinner.getSelectedItem().toString());
        });
        builder.setNegativeButton(R.string.cancel, (dialog, which) -> {

        });
        return builder.create();
    }
}

File: app/src/main/java/android/mahendra/attendancemanager/dialogs/PeriodDialogFragment.kt
package android.mahendra.attendancemanager.dialogs

import android.app.ActionBar
import android.app.AlertDialog
import android.app.Dialog
import android.app.TimePickerDialog
import android.content.Context
import android.content.DialogInterface
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.mahendra.attendancemanager.R
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.widget.*
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.DialogFragment
import java.util.*

class PeriodDialogFragment : DialogFragment() {
    private var subjectTitle: String? = null
    private var allSubjectTitles: List<String> = emptyList()
    private lateinit var toolbar: Toolbar
    private var callbacks: Callbacks? = null

    interface Callbacks {
        val subjectTitles: List<String>
        fun onPeriodSelected(title: String?)
        fun onDeletePeriod(title: String?)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putString(KEY_PERIOD_TITLE, subjectTitle)
        outState.putStringArrayList(KEY_TITLE_OPTIONS, allSubjectTitles as ArrayList<String>)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        try {
            callbacks = context as Callbacks
        } catch (ex: ClassCastException) {
            throw java.lang.ClassCastException("callbacks must be implemented by $context")
        }
    }

    override fun onDetach() {
        super.onDetach()
        callbacks = null
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (savedInstanceState != null) {
            subjectTitle = savedInstanceState.getString(KEY_PERIOD_TITLE, null)
            allSubjectTitles = savedInstanceState.getStringArrayList(KEY_TITLE_OPTIONS) as List<String>
        } else {
            allSubjectTitles = callbacks!!.subjectTitles
            subjectTitle = arguments?.getString(ARG_PERIOD_TITLE, null)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        val view = inflater.inflate(R.layout.dialog_choose_period, container, false)
        toolbar = view.findViewById(R.id.toolbar)
        val removePeriodButton = view.findViewById<Button>(R.id.remove_period_button)
        val periodTimeButton  = view.findViewById<Button>(R.id.period_time_button)
        val spinner = view.findViewById<Spinner>(R.id.spinner_select_period)
        removePeriodButton.setOnClickListener { v1: View? ->
            callbacks!!.onDeletePeriod(subjectTitle)
            Toast.makeText(activity, "Removed $subjectTitle", Toast.LENGTH_SHORT).show()
            dismiss()
        }
        var hours: Int = 0
        var minutes: Int = 30
        periodTimeButton.setOnClickListener {
            val dialog = TimePickerDialog(
                    activity,
                    TimePickerDialog.OnTimeSetListener() { _: TimePicker, _: Int, _: Int ->

                    }, hours, minutes, false)
            dialog.show()
        }


        if (subjectTitle != null) {
            removePeriodButton.visibility = View.VISIBLE

        } else {
            removePeriodButton.visibility = View.GONE
        }
        val adapter: ArrayAdapter<String> = ArrayAdapter(requireActivity(), android.R.layout.simple_spinner_item)
        adapter.addAll(allSubjectTitles)
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        spinner.adapter = adapter

        if (subjectTitle != null) {
            val position = adapter.getPosition(subjectTitle)
            spinner.setSelection(position)
        }
        return view
    }

    override fun getTheme(): Int {
        return R.style.DialogTheme
    }

//    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
//
//
////        val builder = AlertDialog.Builder(activity)
////        builder.setTitle(dialogTitleId)
////        builder.setView(view)
////        builder.setPositiveButton(positiveButtonTextId) { dialog: DialogInterface?, which: Int -> callbacks!!.onPeriodSelected(spinner.selectedItem.toString()) }
////        builder.setNegativeButton(R.string.cancel) { dialog: DialogInterface?, which: Int -> }
////        return builder.create()
//        Log.i(TAG, "on create dialog called")
//        return super.onCreateDialog(savedInstanceState)
//    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        toolbar.title = "Add Period"
        toolbar.inflateMenu(R.menu.period_dialog_fragment)

    }

    companion object {
        private const val TAG = "PeriodDialogFragment"
        private const val ARG_PERIOD_TITLE = "period title"
        private const val KEY_TITLE_OPTIONS = "title options"
        private const val KEY_PERIOD_TITLE = "period_title"
        fun newInstance(periodTitle: String?): PeriodDialogFragment {
            val args = Bundle()
            args.putString(ARG_PERIOD_TITLE, periodTitle)
            val fragment = PeriodDialogFragment()
            fragment.arguments = args
            return fragment
        }
    }
}